R/methods-group-features.R

```diff
@@ -0,0 +1,1049 @@
+                obj_sub <- .filter_file_XCMSnExp(object, sample_idx,
+                                                 keepFeatures = TRUE)
+                ## Can happen that some of the features are not present in the
+                ## subsetted object. Will put them into their own individual
+                ## groups later.
+                idx_miss <- which(!rownames(fvals)[idx] %in%
+                                  rownames(featureDefinitions(obj_sub)))
+                if (length(idx_miss)) {
+                    tmp <- idx[idx_miss]
+                    idx <- idx[-idx_miss]
+                    idx_miss <- tmp
+                }
+                if (length(idx) > 1) {
+                    eics <- do.call(
+                        featureChromatograms,
+                        args = c(list(obj_sub, features = rownames(fvals)[idx],
+                                      filled = TRUE), param@dots))
+                    if (param@onlyPeak)
+                        eics <- removeIntensity(eics,
+                                                which = "outside_chromPeak")
+                    res <- do.call(
+                        .group_eic_similarity,
+                        args = c(list(as(eics, "MChromatograms"),
+                                      aggregationFun = ffun,
+                                      threshold = param@threshold,
+                                      ALIGNFUN = param@ALIGNFUN,
+                                      ALIGNFUNARGS = param@ALIGNFUNARGS,
+                                      FUN = param@FUN,
+                                      FUNARGS = param@FUNARGS,
+                                      groupFun = param@groupFun), param@dots))
+                } else res <- factor(1)
+                f_new[idx] <- paste0(fg, ".", MsFeatures:::.format_id(res))
+                if (length(idx_miss))
+                    f_new[idx_miss] <- paste0(
+                        fg, ".", MsFeatures:::.format_id(
+                                                  seq((length(levels(res)) + 1),
+                                                      length.out = length(
+                                                          idx_miss))))
+            } else
+                f_new[idx] <- paste0(fg, ".001")
+            pb$tick(1)
+        }
+        featureDefinitions(object)$feature_group <- f_new
+        xph <- new("XProcessHistory", param = param, date = date(),
+                   type = .PROCSTEP.FEATURE.GROUPING,
+                   fileIndex = seq_along(fileNames(object)),
+                   msLevel = as.integer(msLevel))
+        object@.processHistory[[(length(object@.processHistory) + 1)]] <- xph
+        validObject(object)
+        object
+    })
+
+#' @title Group EICs based on their correlation
+#'
+#' @description
+#'
+#' `groupEicCorrelation` groups (extracted ion) chromatograms (EICs) based on
+#' their similarity with each other. If this correlation is `>=` than the
```

Comment:
I sometimes contemplate whether we need both the correlation threshold and the p-Value to avoid false positive (high) correlations based on too few scans...

Reply:
totally agree. But what I did **not** want to happen here is that we restrict ourselfs to just Pearson's correlation for similarity calculation (i.e. using `cor.test` instead of `cor` which then calculates also the p-value). I wanted to keep the similarity calculation as customizable as possible: we can use any function that returns a numeric value as similarity function. Instead of a p-value I would then maybe require a minimum number of matched/valid data points?

Reply:
I agree to keep things simple (and generic). So a user that wants a second threshhold (i.e. <= p) needs to implement `correlationP0995()` overwriting any standard values. Perfectly fine.